### PR TITLE
[RPC] Add local rpc node ProgPow GPU mining support

### PR DESCRIPTION
**Pull request by blondfrogs**
- Adding new command line argument "-miningaddress"
- Added new rpc call **pprpcsb** which stands for prog pow rpc submit block

### CODE REVIEW

File: src/rpc/mining.cpp [hunk 1]
```diff
@@ -1035,6 +1140,7 @@ static const CRPCCommand commands[] =
     { "mining",             "prioritisetransaction",  &prioritisetransaction,  {"txid","dummy","fee_delta"} },
     { "mining",             "getblocktemplate",       &getblocktemplate,       {"template_request"} },
     { "mining",             "submitblock",            &submitblock,            {"hexdata","dummy"} },
+    { "mining",             "pprpcsb",                &pprpcsb,            {"header_hash","mix_hash", "nonce"} },
```
Comment (mimirmim) [inline src/rpc/mining.cpp]: Little bit nit picky here but, space after the comma. You can fix it now, or forget about it when we do a code cleanup round.

File: src/rpc/mining.cpp [hunk 2]
```diff
@@ -789,6 +808,92 @@ class submitblock_StateCatcher : public CValidationInterface
...
+        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Block does not solve the boundary");
+
+
+    uint256 hash = blockptr->GetHash();
+    {
+        LOCK(cs_main);
+        const CBlockIndex* pindex = LookupBlockIndex(hash);
+        if (pindex) {
+            if (pindex->IsValid(BLOCK_VALID_SCRIPTS)) {
+                return "duplicate";
```
Comment (mimirmim) [inline src/rpc/mining.cpp]: Maybe do JSONRPC errors..?

File: src/rpc/mining.cpp [hunk 3]
```diff
@@ -789,6 +808,92 @@ class submitblock_StateCatcher : public CValidationInterface
...
+
+    bool new_block;
+    submitblock_StateCatcher sc(blockptr->GetHash());
+    RegisterValidationInterface(&sc);
+    bool accepted = ProcessNewBlock(Params(), blockptr, /* fForceProcessing */ true, /* fNewBlock */ &new_block);
+    UnregisterValidationInterface(&sc);
+    if (!new_block) {
+        if (!accepted) {
+            // TODO Maybe pass down fNewBlock to AcceptBlockHeader, so it is properly set to true in this case?
+            return "invalid";
```
Comment (mimirmim) [inline src/rpc/mining.cpp]: Maybe do JSONRPC errors on these returns?

File: src/rpc/mining.cpp [hunk 4]
```diff
@@ -768,6 +781,12 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
     result.pushKV("accumulatorhashes", mapaccumulatorhashes);
     result.pushKV("proofoffullnodehash", pblock->hashPoFN.GetHex());
 
+    if (pblock->IsProgPow()) {
+        result.pushKV("pprpcheader", pblock->GetProgPowHeaderHash().GetHex());
+        result.pushKV("pprpcepoch", progpow::get_epoch_number(pblock->nHeight));
```
Comment (CaveSpectre11) [inline src/rpc/mining.cpp]: These need to be added to the getblocktemplate Result help. (other stuff needs to be added too, but not necessary for this PR; lots of cleanup is needed in the cli help sections)

File: src/rpc/mining.cpp [hunk 5]
```diff
@@ -789,6 +808,92 @@ class submitblock_StateCatcher : public CValidationInterface
     }
 };
 
+static UniValue pprpcsb(const JSONRPCRequest& request) {
```
Comment (CaveSpectre11) [inline src/rpc/mining.cpp]: I presume the command name is meaningful?

File: src/rpc/mining.cpp [hunk 6]
```diff
@@ -599,15 +601,26 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
...
-        pblocktemplate = BlockAssembler(Params()).CreateNewBlock(scriptDummy, fSupportsSegwit, false, false);
+        // Get mining address if it is set
+        std::string address = gArgs.GetArg("-miningaddress", "");
+        CTxDestination dest =  DecodeDestination(address);
+
+        CScript script;
+        if (IsValidDestination(dest)) {
+            script = GetScriptForDestination(dest);
+        } else {
+            script = CScript() << OP_TRUE;;
```
[the remark this answers is not in the document]
Comment (CaveSpectre11) [reply src/rpc/mining.cpp]: Yes this raised a question in my mind as well.  What is the behavior when an invalid destination is used and what would the user expect?

File: src/rpc/mining.cpp (hunk 5 above)
Comment (blondfrogs) [reply src/rpc/mining.cpp]: progpowrpcsubmitblock

File: src/rpc/mining.cpp (hunk 2 above)
Comment (blondfrogs) [reply src/rpc/mining.cpp]: I am just using the same format that submit block is using.

File: src/rpc/mining.cpp (hunk 3 above)
Comment (blondfrogs) [reply src/rpc/mining.cpp]: I am just using the same format that submit block is using.

File: src/rpc/mining.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (blondfrogs) [reply src/rpc/mining.cpp]: Fixed

File: src/rpc/mining.cpp (hunk 3 above)
Comment (blondfrogs) [reply src/rpc/mining.cpp]: Not going to fix

File: src/rpc/mining.cpp (hunk 4 above)
Comment (blondfrogs) [reply src/rpc/mining.cpp]: Added

File: src/rpc/mining.cpp (hunk 3 above)
Comment (CaveSpectre11) [reply src/rpc/mining.cpp]: I'm okay with this. Since it's prevalent throughout places in the code, it can be done on a separate refactoring PR.

**Comment by CaveSpectre11** [comment]
utACK https://github.com/Veil-Project/veil/pull/753/commits/3b5ffaf0445a2ff3ddf8452fb139a3386101f922

**Pull request merged**